[Bug Report] Disconnection race condition

- **OS and version used:** Android 4.4
- **Java runtime used:** 1.6
- **SDK version used:** Java release 2020-11-16 (1.27.0) and Java release 2021-08-09 (1.32.1)

## Description of the issue
When trying to close the connection, we ran into a race condition with the SDK. The logs below show the race condition between our application successfully closing the connection (status DISCONNECTED) and the SDK restoring the connection (status DISCONNECTED_RETRYING)  

Thread ID 854 in our application called closeNow() to close  the connection but at the same time thread ID 1264 which appear to be from the SDK restored it at the same moment.  

The result is, we tried to close the connection but the SDK restored it so we are still connected. 

The logs are confusing. IotHubTransport set status to DISCONNECTED then DISCONNECTED_RETRYING.
It also print a trace about client being successfully closed but also not being able to disconnect. 

## Code sample exhibiting the issue
No specific code to provide.

## Console log of the issue
```
Timestamp            PID   TID
09-07 18:12:40.661   586   854 I DeviceClient:    Closing device client...
09-07 18:12:40.691   586   854 I IotHubTransport: Updating transport status to new status DISCONNECTED with reason CLIENT_CLOSE
09-07 18:12:40.691   586  1264 W IotHubTransport: Updating transport status to new status DISCONNECTED_RETRYING with reason NO_NETWORK
09-07 18:12:40.691   586  1264 W IotHubTransport: com.microsoft.azure.sdk.iot.device.exceptions.ProtocolException: Mqtt connection lost
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.exceptions.PahoExceptionTranslator.convertToMqttException(PahoExceptionTranslator.java:75)
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.connectionLost(Mqtt.java:389)
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at org.eclipse.paho.client.mqttv3.internal.CommsCallback.connectionLost(CommsCallback.java:304)
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at org.eclipse.paho.client.mqttv3.internal.ClientComms.shutdownConnection(ClientComms.java:441)
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:197)
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at java.lang.Thread.run(Thread.java:841)
09-07 18:12:40.691   586  1264 W IotHubTransport: Caused by: Connection lost (32109) - java.io.EOFException
09-07 18:12:40.691   586  1264 W IotHubTransport: 	... 2 more
09-07 18:12:40.691   586  1264 W IotHubTransport: Caused by: java.io.EOFException
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at java.io.DataInputStream.readByte(DataInputStream.java:77)
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at org.eclipse.paho.client.mqttv3.internal.wire.MqttInputStream.readMqttWireMessage(MqttInputStream.java:92)
09-07 18:12:40.691   586  1264 W IotHubTransport: 	at org.eclipse.paho.client.mqttv3.internal.CommsReceiver.run(CommsReceiver.java:137)
09-07 18:12:40.691   586  1264 W IotHubTransport: 	... 1 more
09-07 18:12:40.691   586   854 I IotHubTransport: Client connection closed successfully
09-07 18:13:40.920   586  1264 W IotHubTransport: Failed to close and re-open the iot hub transport connection, checking if another retry attempt should be made
09-07 18:13:40.920   586  1264 W IotHubTransport: com.microsoft.azure.sdk.iot.device.exceptions.ProtocolException: Unable to disconnect
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.exceptions.PahoExceptionTranslator.convertToMqttException(PahoExceptionTranslator.java:75)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.disconnect(Mqtt.java:166)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.connect(Mqtt.java:128)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttMessaging.start(MqttMessaging.java:57)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttIotHubConnection.open(MqttIotHubConnection.java:202)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.openConnection(IotHubTransport.java:785)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.singleReconnectAttempt(IotHubTransport.java:916)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.reconnect(IotHubTransport.java:865)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.handleDisconnection(IotHubTransport.java:820)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.IotHubTransport.onConnectionLost(IotHubTransport.java:253)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.ReconnectionNotifier$1.run(ReconnectionNotifier.java:21)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at java.lang.Thread.run(Thread.java:841)
09-07 18:13:40.920   586  1264 W IotHubTransport: Caused by: Connect already in progress (32110)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at org.eclipse.paho.client.mqttv3.internal.ClientComms.close(ClientComms.java:237)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.close(MqttAsyncClient.java:1617)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at org.eclipse.paho.client.mqttv3.MqttAsyncClient.close(MqttAsyncClient.java:1605)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.MqttConnection.close(MqttConnection.java:180)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	at com.microsoft.azure.sdk.iot.device.transport.mqtt.Mqtt.disconnect(Mqtt.java:158)
09-07 18:13:40.920   586  1264 W IotHubTransport: 	... 10 more
09-07 18:13:42.021   586  1264 I IotHubTransport: The connection to the IoT Hub has been established
09-07 18:13:42.021   586  1264 I IotHubTransport: Updating transport status to new status CONNECTED with reason CONNECTION_OK

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java
@@ -88,6 +88,7 @@ public class IotHubTransport implements IotHubListener
 
     private ScheduledExecutorService taskScheduler;
 
+    // state lock to prevent simultaneous close and reconnect operations. Also prevents multiple reconnect threads from executing at once
     final private Object reconnectionLock = new Object();
 
     // State lock used to communicate to the IotHubSendTask thread when a message needs to be sent or a callback needs to be invoked.
@@ -103,6 +104,9 @@ public class IotHubTransport implements IotHubListener
     private final ProxySettings proxySettings;
     private SSLContext sslContext;
     private final boolean isMultiplexing;
+
+    // Flag set when close() starts. Acts as a signal to any running reconnection logic to not try again.
+    private boolean isClosing;
 
     // Used to store the CorrelationCallbackMessage for a correlationId
     private final Map<String, CorrelatingMessageCallback> correlationCallbacks = new ConcurrentHashMap<>();
@@ -409,6 +413,8 @@ public class IotHubTransport implements IotHubListener
             throw new TransportException("Open cannot be called while transport is reconnecting");
         }
 
+        this.isClosing = false;
+
         // The default config is only null when someone creates a multiplexing client and opens it before
         // registering any devices to it. No need to check for SAS token expiry if no devices are registered yet.
         if (this.getDefaultConfig() != null)
@@ -487,39 +493,45 @@ public class IotHubTransport implements IotHubListener
             throw new IllegalArgumentException("reason cannot be null");
         }
 
-        this.cancelPendingPackets();
+        this.isClosing = true;
 
-        this.invokeCallbacks();
-
-        if (this.taskScheduler != null)
+        // Wait until no reconnection logic is taking place
+        synchronized (this.reconnectionLock)
         {
-            this.taskScheduler.shutdown();
-        }
+            this.cancelPendingPackets();
 
-        try
-        {
-            if (this.iotHubTransportConnection != null)
+            this.invokeCallbacks();
+
+            if (this.taskScheduler != null)
             {
-                this.iotHubTransportConnection.close();
-            }
-        }
-        finally
-        {
-            this.updateStatus(IotHubConnectionStatus.DISCONNECTED, reason, cause);
-
-            // Notify send thread to finish up so it doesn't survive this close
-            synchronized (this.sendThreadLock)
-            {
-                this.sendThreadLock.notifyAll();
+                this.taskScheduler.shutdown();
             }
 
-            // Notify receive thread to finish up so it doesn't survive this close
-            synchronized (this.receiveThreadLock)
+            try
             {
-                this.receiveThreadLock.notifyAll();
+                if (this.iotHubTransportConnection != null)
+                {
+                    this.iotHubTransportConnection.close();
+                }
             }
+            finally
+            {
+                this.updateStatus(IotHubConnectionStatus.DISCONNECTED, reason, cause);
 
-            log.debug("Client connection closed successfully");
+                // Notify send thread to finish up so it doesn't survive this close
+                synchronized (this.sendThreadLock)
+                {
+                    this.sendThreadLock.notifyAll();
+                }
+
+                // Notify receive thread to finish up so it doesn't survive this close
+                synchronized (this.receiveThreadLock)
+                {
+                    this.receiveThreadLock.notifyAll();
+                }
+
+                log.debug("Client connection closed successfully");
+            }
         }
     }
 
@@ -1218,6 +1230,12 @@ public class IotHubTransport implements IotHubListener
                 && transportException != null
                 && transportException.isRetryable())
         {
+            if (this.isClosing)
+            {
+                log.trace("Abandoning reconnection logic since this client has started closing. Releasing reconnectionLock");
+                return;
+            }
+
             log.trace("Attempting reconnect attempt {}", reconnectionAttempts);
             reconnectionAttempts++;
 

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java
@@ -493,6 +493,10 @@ public class IotHubTransport implements IotHubListener
             throw new IllegalArgumentException("reason cannot be null");
         }
 
+        // Set the flag outside of the synchronized block so that any currently
+        // running reconnection logic knows to give up when this flag is set to true.
+        // Then the rest of the close() code is in the synchronization block so that
+        // it waits for the reconnection logic to end before it starts.
         this.isClosing = true;
 
         // Wait until no reconnection logic is taking place

--- a/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java
+++ b/device/iot-device-client/src/main/java/com/microsoft/azure/sdk/iot/device/transport/IotHubTransport.java
@@ -1232,7 +1232,7 @@ public class IotHubTransport implements IotHubListener
         {
             if (this.isClosing)
             {
-                log.trace("Abandoning reconnection logic since this client has started closing. Releasing reconnectionLock");
+                log.trace("Abandoning reconnection logic since this client has started closing");
                 return;
             }
 

